frontend/src/metabase/routes-stable-id-aware.tsx

```diff
@@ -0,0 +1,147 @@
+import { useEffect, useMemo, useState } from "react";
+import type { WithRouterProps } from "react-router";
+import { match } from "ts-pattern";
+
+import {
+  type BaseEntityId,
+  isBaseEntityID,
+} from "metabase-types/api/entity-id";
+
+import { useTranslateEntityIdQuery } from "./api/entity-id";
+import { NotFound } from "./components/ErrorPages";
+import LoadingAndErrorWrapper from "./components/LoadingAndErrorWrapper";
+
+type ResourceType = "dashboard" | "collection" | "card" | "dashboard-tab";
+type ParamType = "param" | "search";
+
+type ParamConfig = {
```

Comment:
Compared to the original tech plan, I ended up refactoring to a simpler configuration with a single array and a "type" to specify if it's a named parameter (eg `:slug`) or a search parameter (eg: `tab`).
It simplified the logic and the api is still readble imho.